omx/gstomxcamera.c

```diff
@@ -0,0 +1,1214 @@
+      klass->cdata.hacks);
+
+  if (!self->comp)
+    return FALSE;
+
+  if (gst_omx_component_get_state (self->comp,
+          GST_CLOCK_TIME_NONE) != OMX_StateLoaded)
+    return FALSE;
+
+  port_index = klass->cdata.out_port_index;
+  self->outport = gst_omx_component_add_port (self->comp, port_index);
+  if (!self->outport)
+    goto port_failed;
+
+  gst_pad_set_element_private (GST_BASE_SRC_PAD (self), self->outport);
+
+  /* Allocate output buffer pool */
+  pool =
+      gst_omx_buffer_pool_new (GST_ELEMENT (self), self->comp, self->outport);
+  if (!pool)
+    goto pool_failed;
+  self->outpool = pool;
+
+  GST_INFO_OBJECT (self, "Opened component %s", klass->cdata.component_name);
+  return TRUE;
+
+port_failed:
+  {
+    GST_ERROR_OBJECT (self, "Failed to add output port %d", port_index);
+    goto cleanup;
+  }
+pool_failed:
+  {
+    GST_ERROR_OBJECT (self, "Failed to create output pool");
+    goto cleanup;
+  }
+cleanup:
+  {
+    if (self->comp)
+      gst_omx_component_free (self->comp);
+    self->comp = NULL;
+    return FALSE;
+  }
+}
+
+static gboolean
+gst_omx_camera_shutdown (GstOMXCamera * self)
+{
+  OMX_STATETYPE state;
+
+  GST_DEBUG_OBJECT (self, "Shutting down OMX camera");
+
+  state = gst_omx_component_get_state (self->comp, 0);
+  if (state > OMX_StateLoaded || state == OMX_StateInvalid) {
+    /* Changing states to Loaded */
+    if (state > OMX_StateIdle) {
+      gst_omx_component_set_state (self->comp, OMX_StateIdle);
+      gst_omx_component_get_state (self->comp, 5 * GST_SECOND);
+    }
+    gst_omx_component_set_state (self->comp, OMX_StateLoaded);
+
+    /* Deallocating buffers to allow state change */
+    if (self->outpool) {
+      gst_buffer_pool_set_active (self->outpool, FALSE);
+      GST_OMX_BUFFER_POOL (self->outpool)->deactivated = TRUE;
+    }
+    gst_omx_port_deallocate_buffers (self->outport);
+
+    if (state > OMX_StateLoaded)
+      gst_omx_component_get_state (self->comp, 5 * GST_SECOND);
+  }
+
+  return TRUE;
+}
+
+static gboolean
+gst_omx_camera_close (GstOMXCamera * self)
+{
+  GstOMXCameraClass *klass = GST_OMX_CAMERA_GET_CLASS (self);
+
+  GST_DEBUG_OBJECT (self, "Closing OMX camera");
+  if (!gst_omx_camera_shutdown (self))
+    return FALSE;
+
+  /* Free resources and OMX component */
+  if (self->outpool) {
+    gst_object_unref (self->outpool);
+    self->outpool = NULL;
+  }
+
+  if (self->comp) {
+    gst_omx_component_free (self->comp);
+    self->comp = NULL;
+    self->outport = NULL;
+  }
+
+  GST_INFO_OBJECT (self, "Closed component %s", klass->cdata.component_name);
+  return TRUE;
+}
+
+
+static GstStateChangeReturn
+gst_omx_camera_change_state (GstElement * element, GstStateChange transition)
+{
+  GstStateChangeReturn ret = GST_STATE_CHANGE_SUCCESS;
+  GstOMXCamera *self = GST_OMX_CAMERA (element);
+
+  switch (transition) {
+    case GST_STATE_CHANGE_NULL_TO_READY:
+      /* open the component */
+      if (!gst_omx_camera_open (self))
```

Comment:
Why do we do this here and not in the start/stop?

Reply:
Yes, change_state can be avoided. open/close calls are made now through start/stop pair 4e962806a3de137cdc02851f967289f6ef2021b1